core/src/main/java/com/linecorp/armeria/common/stream/DefaultByteStreamMessage.java

```diff
@@ -0,0 +1,376 @@
+        public void onSubscribe(Subscription s) {
+            requireNonNull(s, "s");
+            upstream = s;
+            downstream.onSubscribe(this);
+        }
+
+        @Override
+        public void onNext(HttpData data) {
+            requireNonNull(data, "data");
+            assert buffer == null;
+            if (completed) {
+                data.close();
+                return;
+            }
+
+            inOnNext = true;
+            try {
+                if (!onNext0(data)) {
+                    return;
+                }
+
+                // Use a while-loop to avoid deep recursive calls.
+                // https://github.com/reactive-streams/reactive-streams-jvm#3.3
+                while (buffer != null && demand > 0) {
+                    final HttpData buffer = this.buffer;
+                    this.buffer = null;
+                    if (!onNext0(buffer)) {
+                        break;
+                    }
+                }
+
+                if (completing && buffer == null) {
+                    downstream.onComplete();
+                    return;
+                }
+
+                if (buffer == null && demand > 0) {
+                    requestOneOrCancel();
+                }
+            } finally {
+                inOnNext = false;
+            }
+        }
+
+        /**
+         * Publishes the {@link HttpData} to the downstream.
+         *
+         * @return true if data is published to the downstream.
+         */
+        private boolean onNext0(HttpData data) {
+            if (position >= end) {
+                // Drop tail bytes. Fully received the desired data already.
+                data.close();
+                upstream.cancel();
+                return false;
+            }
+
+            final int dataSize = data.length();
+            final int dataEnd = IntMath.saturatedAdd(position, dataSize);
+            final int skipBytes = Math.max(0, IntMath.saturatedSubtract(offset, position));
+            final int dropBytes = Math.max(0, IntMath.saturatedSubtract(dataEnd, end));
+
+            if (skipBytes >= dataSize) {
+                // Skip the entire data and request the next element.
+                data.close();
+                position += dataSize;
+                requestOneOrCancel();
+                return false;
+            }
+
+            if (skipBytes == 0 && dropBytes == 0) {
+                // Use the data as is.
+                position += dataSize;
+                downstream.onNext(data);
+            } else {
+                try {
+                    final int slicedDataSize = dataSize - skipBytes - dropBytes;
+                    final HttpData slicedData = retainedSlice(data, skipBytes, slicedDataSize);
+                    final int nextOffset = skipBytes + slicedDataSize;
+
+                    if (dropBytes == 0) {
+                        final int remainingLength = dataSize - nextOffset;
+                        if (remainingLength > 0) {
```

Comment:
I'm confused about this condition 😅 
Isn't `remainingLength == dropBytes`, hence this if statement can never pass?

Reply:
Good point. It is cruft. Previously, `ByteStreamMessage` had `bufferSize()` operation. If an `HttpData` is larger than the buffer size, the code was used to slice the remaining buffer.
After we decided to remove `bufferSize()` in `ByteStreamMessage`, I should've completely removed `buffer` field.